### PR TITLE
Implement RoundRobin logic in RoundRobinInetAddressResolver#resolveAll

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Motivation:
Now the ```resolveAll``` method of RoundRobinInetAddressResolver returns results without any rotation and shuffling. As a result, it doesn't force any round-robin for clients that get a result of ```resolveAll``` and use addresses from the result one by one for a connection establishing until success. This commit implements round-robin in RoundRobinInetAddressResolver#resolveAll. These improvements inspired by the discussion here: https://github.com/AsyncHttpClient/async-http-client/issues/1285

Modifications:
Rotate collection from internal ```resolveAll``` call by index, which is incremented every call to RoundRobinInetAddressResolver#resolveAll method.
Random replaced by an incrementing counter, which makes code cheaper and guarantees predictable address order in tests.

Result:
Improved ```RoundRobinInetAddressResolver``` is compatible with clients that use ```resolveAll``` result.

### CODE REVIEW

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 1]
```diff
@@ -52,29 +57,46 @@ protected void doResolve(final String inetHost, final Promise<InetAddress> promi
...
     @Override
-    protected void doResolveAll(String inetHost, Promise<List<InetAddress>> promise) throws Exception {
-        nameResolver.resolveAll(inetHost, promise);
+    protected void doResolveAll(String inetHost, final Promise<List<InetAddress>> promise) throws Exception {
+        nameResolver.resolveAll(inetHost).addListener(new FutureListener<List<InetAddress>>() {
+            @Override
+            public void operationComplete(Future<List<InetAddress>> future) throws Exception {
+                if (future.isSuccess()) {
+                    List<InetAddress> inetAddresses = future.getNow();
+                    if (inetAddresses.size() > 0) {
```
Comment (netkins) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: MAJOR Use isEmpty() to check whether the collection is empty or not. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1155)

File: resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java [hunk 2]
```diff
@@ -36,6 +37,10 @@
 @UnstableApi
 public class RoundRobinInetAddressResolver extends InetNameResolver {
     private final NameResolver<InetAddress> nameResolver;
+    // It should be an AtomicInteger to be fully thread safe.
+    // On the same time, int var is atomic and there is no strict needs in happens-before or memory barrier for this
+    // particular variable, so should be fine to use in a relaxed manner without volatile or AtomicInteger.
+    private int index;
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: Just make it an AtomicInteger please.

Comment (Spikhalskiy) [inline resolver/src/main/java/io/netty/resolver/RoundRobinInetAddressResolver.java]: addressed

**Comment by normanmaurer** [comment]
@Spikhalskiy thanks!

Cherry-picked into 4.1 (5eebe9a06c6873e797b5f8c847086c35ae990ef5)

**Pull request closed without merging**